language-plutus-core/stdlib/Language/PlutusCore/StdLib/Type.hs

```diff
@@ -633,39 +633,96 @@ packPatternFunctorBodyN ann dataName argVars patBodyN = do
         $ TyVar ann rec
 
 -- | Construct a data type out of pieces.
-getTyFix :: FromDataPieces ann (Type TyName ann)
-getTyFix ann name argVars patBodyN = do
+getPackedType :: FromDataPieces ann (Type TyName ann)
+getPackedType ann dataName argVars patBodyN = do
     withSpine <- getWithSpine ann argVars
-    withSpine . TyIFix ann <$> packPatternFunctorBodyN ann name argVars patBodyN
+    withSpine . TyIFix ann <$> packPatternFunctorBodyN ann dataName argVars patBodyN
 
 -- | An auxiliary type for returning a polymorphic @wrap@. Haskell's support for impredicative
 -- polymorphism isn't good enough to do without this.
-newtype PolyWrap ann =
-  PolyWrap { unPolyWrap :: forall term . TermLike term TyName Name
-                        => [Type TyName ann] -> term ann -> term ann
-           }
+newtype PolyWrap ann = PolyWrap
+    (forall term. TermLike term TyName Name => [Type TyName ann] -> term ann -> term ann)
 
 -- | Make a generic @wrap@ that takes a spine of type arguments and the rest of a term, packs
 -- the spine using the CPS trick and passes the spine and the term to 'IWrap' along with a 1-ary
 -- pattern functor constructed from pieces of a data type passed as arguments to 'getWrap'.
-getWrap :: FromDataPieces ann (PolyWrap ann)
-getWrap ann name argVars patBody = do
-    pat1 <- packPatternFunctorBodyN ann name argVars patBody
+getPackedWrap :: FromDataPieces ann (PolyWrap ann)
+getPackedWrap ann dataName argVars patBodyN = do
+    pat1 <- packPatternFunctorBodyN ann dataName argVars patBodyN
     toSpine <- getToSpine ann
     let instVar v ty = TyDecl ann ty $ tyVarDeclKind v
     return $ PolyWrap $ \args ->
         let argVarsLen = length argVars
             argsLen = length args
             in if argVarsLen == argsLen
                 then iWrap ann pat1 . toSpine $ zipWith instVar argVars args
-                else throw . IndicesLengthsMismatchException argVarsLen argsLen $ void name
+                else throw . IndicesLengthsMismatchException argVarsLen argsLen $ void dataName
+
+-- Note that the pattern functor of the resulting type is always the same, it's the index
+-- that differs.
+-- | Construct a 'RecursiveType' by passing a 0-ary pattern functor to 'TyIFix' and 'IWrap'
+-- /as an index/.
+makeRecursiveType0
+    :: ann              -- ^ An annotation placed everywhere we do not have annotations.
+    -> TyName ann       -- ^ The name of the data type being defined.
+    -> Type TyName ann  -- ^ The body of the pattern functor.
+    -> Quote (RecursiveType ann)
+makeRecursiveType0 ann dataName patBody0 = do
+    rec <- freshTyName ann "rec"
+    f   <- freshTyName ann "f"
+    let indKind = KindArrow ann (Type ann) $ Type ann
+        recKind = KindArrow ann indKind $ Type ann
+        pat1
+            = TyLam ann rec recKind
+            . TyLam ann f indKind
+            . TyApp ann (TyVar ann f)
+            . TyApp ann (TyVar ann rec)
+            $ TyVar ann f
```

Comment:
I'd mildly prefer to use the abstractions from `MkPlc` so we can say e.g. 
```
let
    recVar = TyVarDecl ann rec recKind
    fVar = TyVarDecl ann f indKind
    pat1 = mkIterTyLam [recVar, fVar] $ mkIterTyApp (mkTyVar fVar) [ mkTyVar recVar, mkTyVar fVar ]
```
and so on. I think that's easier to read!

Applies throughout.

Reply:
> I think that's easier to read!

It's also wrong :)

Your code means `\rec f -> f rec f`, while it's `\rec f -> f (rec f)` actually.

So the only remaining case is the two lambdas and I'd rather keep the code straight than introduce an indirection given that benefit is really small.

Reply:
I mostly find it easier having variable names and their types/kinds bound together but YMMV :)

Reply:
Do we have something like `freshTyVar`? That would be better than using `freshTyName`, then manually creating a variable, then finally binding it in an expression.

But let's not be blocked on this.

Reply:
Yes, definitely not blocking, and I agree `freshTyVar :: a -> Kind a -> Quote (TyVarDecl tyname a)` would be good.